pkg/ddl/job_worker.go

```diff
@@ -529,19 +530,32 @@ func (w *worker) prepareTxn(job *model.Job) (kv.Transaction, error) {
 // non-zero, caller should wait for other nodes to catch up.
 func (w *worker) transitOneJobStep(
 	jobCtx *jobContext,
-	job *model.Job,
+	jobW *model.JobW,
 	sysTblMgr systable.Manager,
 ) (int64, error) {
-	var (
-		err error
-	)
-
+	failpoint.InjectCall("beforeTransitOneJobStep", jobW)
+	job := jobW.Job
 	txn, err := w.prepareTxn(job)
 	if err != nil {
 		return 0, err
 	}
 	jobCtx.metaMut = meta.NewMutator(txn)
 
+	// we are using optimistic txn in nearly all DDL related transactions, if
+	// time range of another concurrent job updates, such as 'cancel/pause' job
+	// or on owner change, overlap with us, we will report 'write conflict', but
+	// if they don't overlap, we query and check inside our txn to detect the conflict.
+	currBytes, err := sysTblMgr.GetJobBytesByIDWithSe(jobCtx.ctx, w.sess, job.ID)
+	if err != nil {
+		// TODO maybe we can unify where to rollback, they are scatting around.
+		w.sess.Rollback()
+		return 0, err
+	}
+	if !bytes.Equal(currBytes, jobW.Bytes) {
```

Comment:
not sure what's the duration of comparing two 6M bytes.

Reply:
worst case, about 210us
```
goos: darwin
goarch: arm64
pkg: try-out/test/misc
cpu: Apple M1
BenchmarkCompareBytes-8   	    6459	    211581 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	try-out/test/misc	1.408s
```
```go
func BenchmarkCompareBytes(b *testing.B) {
	size := 6 << 20
	l := make([]byte, 0, size)
	for i := 0; i < size; i++ {
		l = append(l, 'a')
	}
	r := make([]byte, 0, size)
	for i := 0; i < size; i++ {
		r = append(r, 'a')
	}

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = bytes.Equal(l, r)
	}
}
```